server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -376,32 +425,43 @@ remoteField
   -> Maybe G.Description
   -> G.ArgumentsDefinition
   -> G.TypeDefinition [G.Name]
-  -> m (FieldParser n ()) -- TODO return something useful, maybe?
+  -> m (FieldParser n RemoteField)
 remoteField sdoc fieldName description argsDefn typeDefn = do
   -- TODO add directives
   argsParser <- argumentsParser argsDefn sdoc
   case typeDefn of
     G.TypeDefinitionObject objTypeDefn -> do
       remoteSchemaObj <- remoteSchemaObject sdoc objTypeDefn
-      pure $ void $ P.subselection fieldName description argsParser remoteSchemaObj
-    G.TypeDefinitionScalar (G.ScalarTypeDefinition desc name' _) ->
-      pure $ P.selection fieldName description argsParser $ remoteFieldScalarParser name' desc
+      let objSelection = P.rawSubselection fieldName description argsParser remoteSchemaObj
+      pure $ objSelection <&> (\(alias, args, _, selSet) ->
+                                 (G.Field alias fieldName (fmap getName <$> args) mempty selSet))
+    G.TypeDefinitionScalar scalarTypeDefn ->
+      let scalarField = remoteFieldScalarParser scalarTypeDefn
+          scalarSelection = P.rawSelection fieldName description argsParser scalarField
+      in
+      pure $ (scalarSelection <&> (\(alias, args, _) -> (G.Field alias fieldName (fmap getName <$> args) mempty [])))
     G.TypeDefinitionEnum enumTypeDefn ->
-      pure $ P.selection fieldName description argsParser $ remoteFieldEnumParser enumTypeDefn
+      let enumField = remoteFieldEnumParser enumTypeDefn
+          enumSelection = P.rawSelection fieldName description argsParser enumField
+      in
+      pure $ enumSelection <&> (\(alias, _, _) -> (G.Field alias fieldName mempty mempty []))
     G.TypeDefinitionInterface ifaceTypeDefn -> do
       remoteSchemaObj <- remoteSchemaInterface sdoc ifaceTypeDefn
-      pure $ void $ P.subselection fieldName description argsParser remoteSchemaObj
+      pure $ P.rawSubselection fieldName description argsParser remoteSchemaObj <&>
+        (\(alias, args, _, selSet) ->
+           (G.Field alias fieldName (fmap getName <$> args) mempty selSet))
     G.TypeDefinitionUnion unionTypeDefn -> do
       remoteSchemaObj <- remoteSchemaUnion sdoc unionTypeDefn
-      pure $ void $ P.subselection fieldName description argsParser remoteSchemaObj
+      pure $ P.rawSubselection fieldName description argsParser remoteSchemaObj <&>
+        (\(alias, args, _, selSet) ->
+           (G.Field alias fieldName (fmap getName <$> args) mempty selSet))
```

Comment:
Visually there's some duplication going on. I understand that the details are different between the various cases. But at least some of them use equal anonymous functions, which we can make into local named definitions, for instance. And we either construct the final `FieldParser` using `rawSelection` or `rawSubselection`. So we can have two locally defined combinators instead of calling `rawSelection` and `rawSubselection` directly, which should clean up the code a little bit.

Reply:
Yes, you're right. I have applied the suggested change.